### PR TITLE
Project week number

### PR DESCRIPTION
**Pull request by apopov77**
The chart option for ability to use project week number in timeline (instead of calendar week number)

### CODE REVIEW

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java [hunk 1]
```diff
@@ -221,6 +241,33 @@ public void languageChanged(Event event) {
...
+      public void taskMoved(@NotNull TaskHierarchyEvent e) {
+        resetLocaleApi();
+      }
+      @Override
+      public void taskRemoved(@NotNull TaskHierarchyEvent e) {
+        resetLocaleApi();
+      }
+      private void resetLocaleApi() {
+        if (myWeekNumberOption.isChecked()) {
+          timeFormatters.setLocaleApi(localeApi);
```
Comment (dbarashev) [inline ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java:266]: This way we will reset cache excessively. I suggest replacing LocaleApi::getWeekNumber(Date) with `getWeekNumberProperty(): SimpleObjectProperty<Function<Date, Integer>>`

Change the function when the option changes, and subscribe to the property changes in the week formatter.

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java [hunk 2]
```diff
@@ -213,6 +218,21 @@ public DateFormat getShortDateFormat() {
...
+        calendar.setTime(date);
+        Integer weekNum = calendar.get(Calendar.WEEK_OF_YEAR);
+
+        if (!myWeekNumberOption.isChecked()) {
+          return weekNum;
+        }
+
+        calendar.setTime(myTaskManager.getProjectStart());
+        Integer startWeekNum = calendar.get(Calendar.WEEK_OF_YEAR);
+        return weekNum - startWeekNum;
```
Comment (dbarashev) [inline ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java:234]: Our users are not programmers, and most likely start numbering with 1. The first project week should be week no 1

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/chart/text/CachingTextFormatter.java [hunk 3]
```diff
@@ -40,7 +41,12 @@ protected TimeUnitText[] getCachedText(Date startDate) {
     return myTextCache.get(startDate);
   }
 
+  public LocaleApi getLocal () {
```
[the remark this answers is not in the document]
Comment (apopov77) [reply biz.ganttproject.core/src/main/java/biz/ganttproject/core/chart/text/CachingTextFormatter.java]: Done

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/chart/text/TimeFormatters.java [hunk 4]
```diff
@@ -88,7 +88,38 @@ public static interface LocaleApi {
...
+      myValue = value;
+      firePropertyChanged(oldValue, value);
+    }
+
+    public void addListener(PropertyChangeListener listener) {
+      myListeners.add(listener);
+    }
+
+    private void firePropertyChanged(T oldValue, T newValue) {
+      for (PropertyChangeListener listener : myListeners) {
```
Comment (dbarashev) [inline biz.ganttproject.core/src/main/java/biz/ganttproject/core/chart/text/TimeFormatters.java:119]: Perhaps it makes sense to use our own listener interface and event class as well. E.g., we can add type arguments to the event class and remove casting. Also, let's move this class into `biz.ganttproject.core.option` and rewrite in Kotlin ;) 

Oh, we have `ChangeValueListener` and `ChangeValueEvent` in `biz.ganttproject.core.option`, albeit without generic types.

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java [hunk 5]
```diff
@@ -190,7 +194,55 @@ public boolean apply(Task input) {
   private final GPOptionGroup myTimelineLabelOptions;
 
   private final FontOption myChartFontOption;
-  private final BooleanOption myWeekNumberOption;
+
+  private final DefaultEnumerationOption myWeekNumberOption;
+  private TimeFormatters.ObservableProperty<java.util.function.Function<Date, Integer>> myWeekNumProperty;
+
+  private final static Function<Date, Integer> US_WEEK_NUMBERING = date -> {
```
Comment (dbarashev) [inline ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java:201]: Let's translate these functions to Kotlin and place in `WeekNumbering.kt` nearby?

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java [hunk 6]
```diff
@@ -242,8 +287,22 @@ public void languageChanged(Event event) {
...
     });
 
-    // clear formatter caches
-    myWeekNumberOption.addChangeValueListener(evt -> {timeFormatters.setLocaleApi(localeApi);});
+    myWeekNumberOption.addChangeValueListener(evt -> {
+      switch (evt.getNewValue().toString()) {
+        case UIConfiguration.WeekOption.US:
+          myWeekNumProperty.setValue(US_WEEK_NUMBERING);
+          break;
+          case UIConfiguration.WeekOption.EUROPEAN:
```
Comment (dbarashev) [inline ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java:295]: something went wrong with indents

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java [hunk 7]
```diff
@@ -190,7 +194,55 @@ public boolean apply(Task input) {
...
+    return localDate.get(weekNumbering.weekOfWeekBasedYear());
+  };
+
+  private final static Function<Date, Integer> EUROPEAN_WEEK_NUMBERING = date -> {
+    WeekFields weekNumbering = WeekFields.of(Locale.UK);
+    LocalDate localDate = date.toInstant().atZone(ZoneOffset.UTC).toLocalDate();
+    return localDate.get(weekNumbering.weekOfWeekBasedYear());
+  };
+
+  private final static Function<Date, Integer> DEFAULD_WEEK_NUMBERING = date -> {
```
Comment (dbarashev) [inline ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java:213]:
```suggestion
  private final static Function<Date, Integer> DEFAULT_WEEK_NUMBERING = date -> {
```

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java (hunk 1 above)
Comment (apopov77) [reply ganttproject/src/main/java/net/sourceforge/ganttproject/chart/ChartModelBase.java]: Observable property class has been created

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package biz.ganttproject.core.option
+
+
+class ObservableProperty<T>(val name: String, private var initValue: T) {
+
+  private val myListeners: MutableList<ChangeValueListener?> = mutableListOf()
```
Comment (dbarashev) [inline biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt:6]:
```suggestion
  private val listeners: MutableList<ChangeValueListener> = mutableListOf()
```

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      firePropertyChanged(oldValue, newValue)
+    }
+
+  fun addListener(listener: ChangeValueListener?) {
+    myListeners.add(listener)
+  }
+
+  private fun firePropertyChanged(oldValue: T, newValue: T) {
+    for (listener in myListeners) {
+      listener!!.changeValue(ChangeValueEvent(name, oldValue, newValue, this))
```
Comment (dbarashev) [inline biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt:21]:
```suggestion
      listener.changeValue(ChangeValueEvent(name, oldValue, newValue, this))
```

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package biz.ganttproject.core.option
+
+
+class ObservableProperty<T>(val name: String, private var initValue: T) {
```
Comment (dbarashev) [inline biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt:4]:
```suggestion
class ObservableProperty<T>(val name: String, initValue: T) {
```

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import java.time.temporal.WeekFields
+import java.util.*
+
+class WeekNumbering {
+
+
+  companion object {
+
+    @JvmStatic
+    val us = Function { date: Date ->
```
Comment (dbarashev) [inline ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt:15]: you can move these objects out of the class to the file level. You can access them in Java using `WeekNumberingKt.getUsNumbering()` (provided that the object is named `usNumbering`)

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      calendar.time = myStartProjectDate
+      val startWeekNum = calendar[Calendar.WEEK_OF_YEAR]
+      weekNum = weekNum - startWeekNum
+      if (weekNum >= 0) {
+        weekNum++
+      }
+      return weekNum
+    }
+
+    override fun equals(`object`: Any?): Boolean {
```
Comment (dbarashev) [inline ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt:51]: it appears that this implementation of equals is the same as the default one, so you can just remove it

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt [hunk 13]
```diff
@@ -0,0 +1,55 @@
+package net.sourceforge.ganttproject.chart
```
Comment (dbarashev) [inline ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt:1]: pls add copyright notice

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt [hunk 14]
```diff
@@ -0,0 +1,24 @@
+package biz.ganttproject.core.option
```
Comment (dbarashev) [inline biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt:1]: pls add a copyright notice

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt (hunk 9 above)
Comment (apopov77) [reply biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt]: Done

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt (hunk 10 above)
Comment (apopov77) [reply biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt]: Done

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt (hunk 11 above)
Comment (apopov77) [reply ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt]: Done

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt (hunk 12 above)
Comment (apopov77) [reply ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt]: Done

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt (hunk 13 above)
Comment (apopov77) [reply ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt:20]: Done

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt (hunk 14 above)
Comment (apopov77) [reply biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt:20]: Done

**Review by dbarashev — APPROVED** [review]
LGTM.
This pull request is good to merge (see some comments though). However, we need to do something with the weeks out of project boundaries in project-relative numbering scheme. It is okay to change them in the next pull request.

File: biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt [hunk 15]
```diff
@@ -1,9 +1,31 @@
+/*
+GanttProject is an opensource project management tool.
+Copyright (C) 2011 Dmitry Barashev, GanttProject Team
```
[the remark this answers is not in the document]
Comment (apopov77) [reply biz.ganttproject.core/src/main/java/biz/ganttproject/core/option/ObservableProperty.kt]: Done

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt [hunk 16]
```diff
@@ -6,50 +25,39 @@ import java.time.ZoneOffset
...
-    }
+val default = Function { date: Date ->
+  val calendar = CalendarFactory.newCalendar()
+  calendar.time = date
+  calendar[Calendar.WEEK_OF_YEAR]
+}
 
-    override fun equals(`object`: Any?): Boolean {
-      return this == `object`
+class RelativeWeekNumbering(private val myStartProjectDate: Date) : Function<Date, Int> {
```
[the remark this answers is not in the document]
Comment (apopov77) [reply ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt]: Done

File: ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt [hunk 17]
```diff
@@ -6,50 +25,39 @@ import java.time.ZoneOffset
...
-    }
-
-    @JvmStatic
-    val default = Function { date: Date ->
-      val calendar = CalendarFactory.newCalendar()
-      calendar.time = date
-      calendar[Calendar.WEEK_OF_YEAR]
-    }
-  }
+val us = Function { date: Date ->
```
[the remark this answers is not in the document]
Comment (apopov77) [reply ganttproject/src/main/java/net/sourceforge/ganttproject/chart/WeekNumbering.kt]: Done

**Pull request merged**